plutus-core/plutus-core/src/PlutusCore/Flat.hs

```diff
@@ -334,25 +337,34 @@ instance Flat Special
 -- See Note [Index (Word64) (de)serialized through Natural]
 instance Flat Index where
     -- encode from word64 to natural
-    encode = encode @Natural . fromIntegral
+    encode =
+#if WORD_SIZE_IN_BITS == 64
+        encode . wordToNatural . fromIntegral @Index @Word
+#else
+        encode @Natural . fromIntegral
+#endif
     -- decode from natural to word64
     decode = do
-        n <- decode
-        case naturalToWord64Maybe n of
-            Just w  -> pure $ Index w
-            Nothing -> fail $ "Index outside representable range: " ++ show n
+        n <- decode @Natural
+#if WORD_SIZE_IN_BITS == 64
+        case naturalToWordMaybe n of
+            Just w  -> pure $ fromIntegral @Word @Index w
+            Nothing ->
+#else
+        if n <= fromIntegral @Word64 @Natural maxBound
+        then pure $ fromIntegral @Natural @Index n
+        else
+#endif
+             fail $ "Index outside representable range: " ++ show n
     -- to be exact, we must not let this be generically derived,
     -- because the `gsize` would derive the size of the underlying Word64,
     -- whereas we want the size of Natural
-    size = sNatural . fromIntegral
-
--- | Downcasts a 'Natural' to a 'Word64', returning 'Nothing' if it won't fit.
--- Not very efficient, could likely be improved.
-naturalToWord64Maybe :: Natural -> Maybe Word64
-naturalToWord64Maybe n =
-    let maxWord64 :: Integer = fromIntegral (maxBound :: Word64)
-        nAsI :: Integer = fromIntegral n
-    in if nAsI <= maxWord64 then Just (fromIntegral nAsI) else Nothing
+    size =
```

Comment:
`size` is only needed when encoding so I also think we don't need this to be fast.